### PR TITLE
haskell.ghc910: overrides for cabal-install and haskell-language-server.

### PR DESCRIPTION
**Pull request by SamirTalwar**
## Description of changes

This adds package overrides for GHC 9.10.1 to make `cabal-install` compile.

It also gets `haskell-language-server` much closer, but unfortunately there's still a way to go. Right now I am blocked on no GHC 9.10.x-compatible version of `apply-refact`.

Apologies if this steps on anyone's toes or is not welcome. It was mostly born of frustration and a "how long can this possibly take?" attitude.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-ghc-9.10.x.nix [hunk 1]
```diff
@@ -14,7 +16,7 @@ self: super: {
   base = null;
   binary = null;
   bytestring = null;
-  Cabal = null;
+  Cabal = doDistribute self.Cabal_3_12_1_0; # GHC ships with Cabal 3.12.0.0, but there is no corresponding cabal-install
```
Comment (maralorn) [inline pkgs/development/haskell-modules/configuration-ghc-9.10.x.nix:19]: We shouldn’t override Cabal globally but only for those packages which can’t build with 3.12.0.

cabal-install already has an override to use Cabal_3_12_1_0 in configuration-common.nix, I think. Why isn’t that enough?

**Comment by maralorn** [comment]
Thanks this is much appreciated and this looks already quite nice.

File: pkgs/development/haskell-modules/configuration-ghc-9.10.x.nix (hunk 1 above)
Comment (SamirTalwar) [reply pkgs/development/haskell-modules/configuration-ghc-9.10.x.nix:19]: Oh, that's a much better idea. I needed it for `cabal-install-solver`.

Looks like these overrides in `configuration-common.nix` only apply if we're working with GHC <= 9.10. When I apply them to GHC 9.10.1 too, it seems to fix the issue so we don't need this `Cabal` override.

Any objection to me removing this line?

https://github.com/NixOS/nixpkgs/blob/5de1564aed415bf9d0f281461babc2d101dd49ff/pkgs/development/haskell-modules/configuration-common.nix#L55

Comment (SamirTalwar) [reply pkgs/development/haskell-modules/configuration-ghc-9.10.x.nix]: I changed this line to refer to `self.ghc.version "9.10.2"`.

**Comment by eldritch-cookie** [comment]
really cool, kinda sad that we duplicated so much work but oh well.
apply-refact is only needed for the hlint plugin, which according to https://haskell-language-server.readthedocs.io/en/stable/support/plugin-support.html is not compilable, i have a devShell which builds HLS [here](https://github.com/eldritch-cookie/network-quick-bread/blob/master/flake.nix). 
By default HLS won't compile with ghc-exactprint 1.10.0.0 i don't remember if jailbreaking works but using 1.9.0.0 definitely does. i was unable to get either the stylish-haskell or the retrie plugin to compile. hope this helped

**Comment by maralorn** [comment]
I don’t think any of those overrides can cause any harm. If they are not optimal, let’s optimize them later. I am gonna merge this now so that other people can iterate on this.

**Comment by SamirTalwar** [comment]
Thanks! At least we’re a bit closer.

@eldritch-cookie, thanks for sharing that. I will be stealing parts of it, I’m sure.

**Pull request merged**